[ML] Anomaly detection for multiple bucket features

tveasey opened the pull request:
Currently, we perform anomaly detection only on a single bucket feature, typically derived from  some aggregation applied to all the documents which fall in that bucket's time interval. However, In many cases interesting events span multiple buckets.

The most important case is where any given bucket is not especially unusual, given the variation w.r.t. our predictions we've historically seen, but a collection of contiguous buckets are jointly unusual.

Such cases can be detected by modelling and detecting unusual events w.r.t. collective properties of a number of contiguous buckets, i.e. features derived from multiple buckets; this PR uses the cumulative prediction error in a sliding window ending at the current bucket. This has proved effective for picking up the sort of events we want to detect, i.e. extended periods where are model is systematically in error.

This constitutes a reasonably significant change to results: increasing the relative importance we'll report for longer lasting anomalies. However, it is giving consistently better area under the P-R curve against our internal test suit. In addition, we pay some extra memory costs (both to compute the feature and for the extra model). This has been more than recouped in other preparatory changes, i.e. #127, #146, etc.

hendrikmuhs commented on bin/autodetect/CCmdLineParser.cc:
```diff
@@ -118,7 +117,7 @@ bool CCmdLineParser::parse(int argc,
                         "Optional flag to enable multi-variate analysis of correlated by fields")
             ("multipleBucketspans",  boost::program_options::value<std::string>(),
-                        "Optional comma-separated list of additional bucketspans - must be direct multiples of the main bucketspan")
+                        "Deprecated - ignored")
```
what's the reason to keep it? Are there clients using this?

droberts195 replied:
I'm happy to completely remove this, although it does then commit us to removing the corresponding options on the Java side.  @dimitris-athanasiou said he would do this.  The functionality to silently drop any settings related to this functionality (which was never documented and not fully tested) will then live in the Java code.

dimitris-athanasiou replied:
I think it's fine to drop everything from the c++ side. I'll prepare the java side and we will merge them both in `6.5`.

dimitris-athanasiou replied:
I raised https://github.com/elastic/elasticsearch/pull/32496. @tveasey, this means you can completely remove the `multipleBucketspans` param in this PR.

tveasey replied:
Ok great. I'll tidy.

hendrikmuhs commented on include/maths/CTimeSeriesBulkFeatures.h:
```diff
@@ -0,0 +1,163 @@
+#include <boost/bind.hpp>
+#include <boost/circular_buffer.hpp>
+#include <boost/iterator/counting_iterator.hpp>
+
```
does not seem to be used in this place

tveasey replied:
I think this was left over from earlier versions, I'll tidy.

hendrikmuhs commented on lib/maths/CTimeSeriesModel.cc:
```diff
@@ -701,68 +737,69 @@ CUnivariateTimeSeriesModel::addSamples(const CModelAddSamplesParams& params,
 
     if (randomSample) {
-        m_SlidingWindow.push_back({randomSample->first, randomSample->second});
+        m_RecentSamples.push_back({randomSample->first, randomSample->second});
```
could be an emplace_back?

tveasey replied:
This is a boost circular buffer which doesn't support emplace_back unfortunately.

hendrikmuhs commented:
I had a first scan over it, looks good.

One remark with respect to naming: I find `multibucket` more descriptive than `bulk`. I think it would be easier of the code would also use `multibucket`.

tveasey commented on include/maths/CTimeSeriesBulkFeatures.h:
```diff
@@ -0,0 +1,163 @@
+    //! Univariate implementation returns zero.
+    template<typename T>
+    static double conformable(const T& /*x*/, double value) {
+        return value;
```
Indeed, comment is out of date.

tveasey commented:
@hendrikmuhs and @dimitris-athanasiou I addressed all your review comments. I also refined aggregation to take account of correlation between the multi-bucket and bucket features, in [this](https://github.com/elastic/ml-cpp/pull/175/commits/3937f40584a9c3ae670bb81b7f2edf80d5ea0a7a) commit. Can you take another look please?

hendrikmuhs commented on include/maths/CTimeSeriesMultibucketFeatures.h:
```diff
@@ -0,0 +1,173 @@
+//! time series from a user's perspective.
+class CTimeSeriesMultibucketFeatures {
+public:
+    //! The geometric weight applied the window.
```
nit: "... applied **to** the window" ?

hendrikmuhs commented on include/model/CAnomalyDetectorModelConfig.h:
```diff
@@ -162,6 +162,10 @@ class MODEL_EXPORT CAnomalyDetectorModelConfig {
 
+    //! The default number of time buckets used to generate multibucket features
+    //! for anomaly detection.
+    static const std::size_t MULTIBUCKET_FEATURE_WINDOW_LENGTH;
```
nit: `MULTIBUCKET_FEATURE_WINDOW_LENGTH` in other places it is `MULTIBUCKET_FEATURES_WINDOW_LENGTH`, feature - > features

hendrikmuhs commented on lib/maths/CTimeSeriesModel.cc:
```diff
@@ -132,14 +168,16 @@ const std::string IS_NON_NEGATIVE_OLD_TAG{"g"};
-const std::string MEAN_ERROR_TAG{"a"};
-const std::string ANOMALIES_TAG{"b"};
-const std::string ANOMALY_FEATURE_MODEL_TAG{"d"};
+// Version 6.4
```
up to 6.5?

hendrikmuhs commented on include/maths/CTimeSeriesMultibucketFeatures.h:
```diff
@@ -0,0 +1,173 @@
+//! anomaly detection. Specifically, unusual values of certain properties
+//! of extended time ranges are often the most interesting events in a
+//! time series from a user's perspective.
+class CTimeSeriesMultibucketFeatures {
```
nit: final?

hendrikmuhs commented on lib/maths/unittest/CTimeSeriesMultibucketFeaturesTest.cc:
```diff
@@ -0,0 +1,131 @@
+void CTimeSeriesMultibucketFeaturesTest::testMean() {
+    // Test we get the value and weight we expect.
+
+    LOG_DEBUG(<< "Univariate");
```
nit: why not make 2 tests?

hendrikmuhs commented on lib/maths/unittest/CTimeSeriesMultibucketFeaturesTest.h:
```diff
@@ -0,0 +1,20 @@
+class CTimeSeriesMultibucketFeaturesTest : public CppUnit::TestFixture {
+public:
+    void testMean();
+    void testContrast();
```
not used

hendrikmuhs commented:
I had another pass, only nitpicks except the version tags which I think should be bumped to 6.5.
